### DISCUSSION TITLE
How to hide or encrypt code during deployment

### DISCUSSION THREAD

**Discussion by JunMa11** [Project-MONAI/monai-deploy-app-sdk#314] [General]
Dear MONAI Team,

I'm really excited about the great tool. It provides a very user-friendly framework to package the AI models.

The framework is built on docker. So the code is visible by entering the docker via interactive mode `-it`.

I'm wondering is there any way to hide or encrypt the code during deployment. E.g,. when sharing the built app with users, we do not hope the users can access the code inside the docker.

**Comment by MMelQin** [comment]
Hi @JunMa11 It is a good question, and valid concern when you have application code that needs to be protected. For MONAI Deploy App SDK, and MONAI in general, as it is open source, the source code is in the open and everyone is welcome to contribute to the development. So the source code is in the open.

For custom code developed using MONAI, it may be obfuscated by compiling the code into pyc, or use PyInstaller as a post package step?

Would like to hear what you think.

**Comment by ericspod** [comment]
If you're concerned with keeping your code or model data hidden from users you've distributed the Docker image to, my understanding is that's it would be really hard to hide what the image contains. The Docker runtime needs to read this data to be able to run the container of course, so at some point even if encrypted it needs to be decrypted using a key your users would have access to somehow. What information there is out there on these sort of security questions are mostly for deploying within cloud environments and rely on complex mechanisms such as TPM. I feel this is beyond our scope to address at the moment.

**Comment by JunMa11** [comment]
Thanks for your replies very much. I agree that it's hard to hide the contents (especially for the code and trained models) inside the docker image when sharing with others.
Pyinstaller and [nuitka](https://nuitka.net/doc/user-manual.html#usage) could be potential solutions, but they're out of the scope of MONAI.

Anyway, thanks for developing this awesome SDK:)
